Automattic/jetpack
# Premium Content / Subscriptions: introduce Paid Newsletter functionality

## Ovearching project

The purpose of this branch is to introduce **paid newsletters functionality** to WordPress.com and possibly Jetpack, at a later time. This mostly connects the Memberships module to the Subscriptions module.

You can read more about this work in pbOQVh-2t2-p2

## Purpose of this Pull request

**This branch should have no functional effect in Jetpack.** The feature is gated by `JETPACK_BETA_BLOCKS` flag and JavaScript parts are gated by `paid-newsletters-in-subscriptions` empty extension.

## Changes proposed in this PR


### WP-Admin: Prompt user to create / manage paid plans from the subscribe block:

<img width="1017" alt="Zrzut ekranu 2022-09-27 o 09 31 58" src="https://user-images.githubusercontent.com/3775068/192462696-1214cef9-ff1f-4850-b4d0-089c6c85b710.png">

### WP-Admin: Ability to mark any post as "Paid Subscriber Only" or "Blog Subscribers Only"


![Zrzut ekranu 2022-11-1 o 17 27 34](https://user-images.githubusercontent.com/3775068/199285381-a960c66c-ea40-41a6-86ad-5f403e3aba1d.png)

This functionality will be presented in every blog post on .com and will serve to help paid features discoverability


### Frontend: Open checkout in the subscribe block

After any newsletter-enabled plan is connected to a site, it will open a checkout option before subscribing user:

<img width="1054" alt="Zrzut ekranu 2022-09-27 o 09 35 08" src="https://user-images.githubusercontent.com/3775068/192463200-ef328695-3ddf-42ab-a0b9-8c092564abd4.png">


### Frontend Block and/or unlock access based on JWT token passed from the checkout modal

On the frontend, we **will block access and prompt to subscribe** in following scenarios:

- Post access in post meta `_jetpack_newsletter_access` is set to "All subscribers", user does not have a valid JWT token with `blog_sub: active` property and is not logged in as post editor
- Post access in post meta `_jetpack_newsletter_access` is set to "Paid subscribers", user does not have a valid JWT token with active subscriptions and is not logged in as post editor

![Zrzut ekranu 2022-11-1 o 17 36 52](https://user-images.githubusercontent.com/3775068/199287359-da4adb0d-a3d3-40e8-bf2a-c6e818fa61f6.png)

The complete set of behaviours is described in pbOQVh-2wm-p2

## Testing instructions

For any payment, you want to use card `4242 4242 4242 4242 4242` and any month/year and any cvc combination.

1. Get a Jetpack site. You can use Jurassic Ninja for example: jurassic.ninja/create/?jetpack-beta&branches.jetpack=paid-newsletters
2. Apply this PR
3. Without enabling `JETPACK_BETA_BLOCKS`, this should have no functional changes. 
    - Please put the subscribe block in a post and test the subscription works.
4. ENable  `JETPACK_BETA_BLOCKS`
    -  either via constant in wp-config.php: `define( 'JETPACK_BETA_BLOCKS', true  );`
    - Or if you are on a Jurassic Ninja site, you can set it in UI: /wp-admin/options-general.php?page=companion_settings
5. To sandbox Store:
    - 1. put `define( 'JETPACK__SANDBOX_DOMAIN', implode( '.', [ 'yourwpcomusername', 'dev', 'dfw', 'wordpress', 'com' ] );` in `wp-config.php` in **Jetpack** site. This will point Jetpack to use your sandbox as a sync host
    - 2. put `define( 'USE_STORE_SANDBOX', true );` in your `0-sandbox.php` in your sandbox. This will switch payments to the test cards and you will be able to test payments.
   - 2. Or if you are on a Jurassic Ninja site, you can set it in UI: /wp-admin/options-general.php?page=companion_settings and add your sandbox URL without protocol (see above)  in JETPACK__SANDBOX_DOMAIN
    - 3. Sandbox `public-api.wordpress.com`
6. Prepare your site:
    1. Connect your site to WPCOM using JT or NGROK
    2. Connect your site to Jetpack
    5. Upgrade to a full plan, using the test card details 
    6. Enable setting at the bottom of /wp-admin/admin.php?page=jetpack#/discussion to `Let visitors subscribe to new posts and comments via email` - after that you will be able to see the subscribe block
8. Testing Subscribe block - wp-admin
    1. Insert the subscribe block to a post
    2. Publish it
    3. Go to the post in incognito, submit the form, and see it behave normally - **there should be no functional change**
    4. . You should see "Add a plan" in the subscribe block menubar (in gutenberg). Click it to set up a plan
9. Setting up a plan
    1. It will redirect you to Calypso, where you should be prompted to set up the Stripe connection. If you are sandboxing store, you will see the button to "skip it" - more info in paYJgx-1IE-p2
    2. Once you will have Stripe enabled, you will need to click the link to add a plan again.
    3. Add a plan, make sure the tab with "Newsletter" has enabled the option to subscribe users checking out to the newsletter.
10. Testing subscribe block - frontend
    1. Go on the site frontend in incognito
    2. Put different email than before in the subscribe block, submit
    3. You should see a checkout window with a plan selection after submitting the subscribe form now
    4. Choose a free option, see the checkout window close and the subscribe flow continue without a functional change
    5. Open the subscribe block in incognito AGAIN and insert another email
    6. Choose a paid plan this time, continue with checkout using test card details
    7. After completing everything, the window should close
    8. You should now be a site subscriber. You will be able to see those in https://wordpress.com/people/followers/yoursitedomain
11. Testing post-level controls for a post set to "paid subscribers"
    1. Go to post
    2. Click in the "Newsletter" panel
    3. Select visibility for Paid Subscribers only
    4. Publish a post
    5. Open an incognito window, log in as the user from 10.6 or continue with the checkout flow
    6. Go to a post in incognito, put the email as the user from 10.6
    7. The modal should open, recognize that the user is a paid subscriber and show thank you screen
    8. Click Thank you
    9. The modal should close,
    10. The page should reload and unlock he content
12. Testing post-level controls for a post set to "paid subscribers"
    1. Go to post
    2. Click in the "Newsletter" panel
    3. Select visibility for "All subscribers"
    4. Publish a post
    5. Open an incognito window, log in as the user from 10.4 or continue subscribing (continuing subscribing may be flawed and described in https://github.com/orgs/Automattic/projects/188#card-86514692 )
    6. Go to a post in incognito, put the email as the user from 10.4
    7. The modal should open, recognize that the user is a subscriber and show thank you screen
    8. Click Thank you
    9. The modal should close,
    10. The page should reload and unlock the content

#### Does this pull request change what data or activity we track or use?
No additional information.
